device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/CorrelatingMessageCallback.java

```diff
@@ -0,0 +1,60 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+package com.microsoft.azure.sdk.iot.device;
+
+import com.microsoft.azure.sdk.iot.device.transport.IotHubTransportPacket;
+
+/**
+ * This interface has a number of call backs to notify when a message is in a specific part of the {@code IotHubTransport} lifecycle.
+ */
+public interface CorrelatingMessageCallback
+{
+
+
+    IotHubClientProtocol protocol = null;
+
+    /**
+     * Executed when the message is queued to the transport.
+     * @param message The message queued to the trasnport.
+     * @param packet The resulting transport packet which give access to the event callback.
+     * @param callbackContext The context sent with the message.
+     */
+    void onQueueRequest(Message message, IotHubTransportPacket packet, Object callbackContext);
+
+    /**
+     * Executed when the message request is queued by the transport.
+     * @param message The message queued to the trasnport.
+     * @param packet The resulting transport packet which give access to the event callback.
+     * @param callbackContext The context sent with the message.
+     */
+    void onSendRequest(Message message, IotHubTransportPacket packet, Object callbackContext);
+
+    /**
+     * Executed when the message request has been sent and IoT Hub has acknowledged the send.
+     * @param packet The message queued to the trasnport.
+     * @param callbackContext The context sent with the message.
+     */
+    void onAcknowledgeSendRequestPacket(IotHubTransportPacket packet, Object callbackContext, Throwable e);
+
+    /**
+     * Executed when a response to the message has been sent by IoT Hub and is being acknowledged by the transport.
+     * @param message The message queued to the transport.
+     * @param callbackContext The context sent with the message.
+     */
+    void onAcknowledgeResponse(Message message, Object callbackContext, Throwable e);
+
+    /**
+     * Executed when a response to the message has been sent by IoT Hub and is being receieved by the transport.
```

Comment:
The style guidance from the docs team is to use lowercased "hub" when referring to hub instances. We only use "IoT Hub" when referencing the Azure service, and even then they want it to be like once at the beginning.